Controller/FolderController.php

```diff
@@ -0,0 +1,159 @@
+    }
+
+    /**
+     * Displays the list of article
+     *
+     * @param \eZ\Publish\API\Repository\Values\Content\Location $location of a folder
+     * @return \Symfony\Component\HttpFoundation\Response
+     */
+    public function showFolderListAction( Location $location )
+    {
+        $response = new Response();
+
+        if ( $location->invisible )
+        {
+            throw new NotFoundHttpException( "Location #$location->id cannot be displayed as it is flagged as invisible." );
+        }
+
+        $response->setSharedMaxAge( $this->getConfigResolver()->getParameter( 'content.default_ttl' ) );
+        $response->headers->set( 'X-Location-Id', $location->id );
+        $response->setVary( 'X-User-Hash' );
+
+        $content = $this->getRepository()
+            ->getContentService()
+            ->loadContentByContentInfo( $location->getContentInfo() );
+
+        // Getting language for the current siteaccess
+        $languages = $this->getConfigResolver()->getParameter( 'languages' );
+
+        $excludedContentTypes = $this->container->getParameter( 'ezdemo.folder.folder_view.excluded_content_types' );
+
+        // Using the criteria helper (a demobundle custom service) to generate our query's criteria.
+        // This is a good practice in order to have less code in your controller.
+        $criteria = $this->get( 'ezdemo.criteria_helper' )->generateListFolderCriterion(
+            $location, $excludedContentTypes, $languages
+        );
+
+        // Generating query
+        $query = new LocationQuery();
+        $query->criterion = $criteria;
+        $query->sortClauses = array(
+            new SortClause\DatePublished()
+        );
+
+        // Initialize pagination.
+        $pager = new Pagerfanta(
+            new ContentSearchAdapter( $query, $this->getRepository()->getSearchService() )
+        );
+
+        $pager->setMaxPerPage( $this->container->getParameter( 'ezdemo.folder.folder_list.limit' ) );
+        $pager->setCurrentPage( $this->getRequest()->get( 'page', 1 ) );
+
+        $includedContentTypeIdentifiers = $this->container->getParameter( 'ezdemo.folder.folder_tree.included_content_types' );
+
+        // Get sub folder structure
+        $subContentCriteria = $this->get( 'ezdemo.criteria_helper' )->generateSubContentCriterion(
+            $location, $includedContentTypeIdentifiers, $languages
+        );
+
+        $subContentQuery = new LocationQuery();
+        $subContentQuery->criterion = $subContentCriteria;
+        $subContentQuery->sortClauses = array(
+            new SortClause\ContentName()
+        );
+
+        $searchService = $this->getRepository()->getSearchService();
+        $subContent = $searchService->findLocations( $subContentQuery );
+
+        $treeItems = array();
+        foreach ( $subContent->searchHits as $hit )
+        {
+            $treeItems[] = $hit->valueObject;
+        }
+
+        return $this->render(
+            'eZDemoBundle:full:folder.html.twig',
```

Comment:
Reason is that this will avoid issues with preview for instance, as template selection will be based on the siteaccess settings.

Note that the remark is also valid for the action above this one.